spec/requests/halls_controller_spec.rb

```diff
@@ -0,0 +1,174 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+
+RSpec.describe HallsController do
+  let(:customer) { create(:user) }
+  let(:manager) { create(:user, :manager) }
+
+  shared_examples 'unauthorized user' do
+    it 'redirects to the root path' do
+      expect(response).to redirect_to(root_path)
+    end
+
+    it 'adds flash message' do
+      expect(flash[:alert]).to eq('You are not authorized to perform this action.')
+    end
+  end
+
+  describe '#index' do
+    subject(:request) { get('/halls') }
+
+    it 'returns http status ok' do
+      request
+      expect(response).to have_http_status(:ok)
+    end
+  end
+
+  describe '#show' do
+    subject(:request) { get("/halls/#{id}") }
+
+    let(:hall) { create(:hall) }
+    let(:id) { hall.id }
+
+    it 'returns http status ok' do
+      request
+      expect(response).to have_http_status(:ok)
+    end
+  end
+
+  describe '#new' do
+    subject(:request) { get('/halls/new') }
+
+    context 'when user is a customer' do
+      before do
+        sign_in(customer)
+        request
```

Comment:
perhaps `request` could be in `shared_examples`? 🤔 because we always do `request` in `it` block